### DISCUSSION TITLE
Type-safe route parameters?

### DISCUSSION THREAD

**Discussion by joelclermont** [larastan/larastan#1632] [Q&A]
I'm trying to eliminate usage of the inline var override anywhere I can.

One place I am stuck is with accessing route parameters. For example if I'm use route model binding like `/companies/{company}` and I want to access `$this->route('company')` where `$this` is a `Request`, is there any way to do that so that it knows it's getting a `Company` model? Right now I'm just inlining `@var Company $company` to accomplish it.

**Comment by szepeviktor** [comment]
> Right now I'm just inlining `@var Company $company` to accomplish it.

Thank you for opening this discussion.
It seems there no solution for this.

**Reply by joelclermont** [reply]
If it can't be made type-safe automatically, is there possibly a way to do it manually with a stub file? Maybe specify a list of all route parameter names and conditionally return the correct model? Has anyone tried something like this?

**Reply by szepeviktor** [reply]
If you were Larastan how would you calculate that the return value is of type `Company`?

**Reply by joelclermont** [reply]
I was thinking something manual I would do in my project. Literally mapping route parameter names to models. 

If I was trying to calculate it, I think it would need more context of route resolution to see which types need to get injected to the controller action, but I'd have to think on that a lot more. I'm happy with a manual stub solution for now if you can point me in the right direction. Otherwise, I'll probably experiment on this later and report back. Thanks!

**Reply by szepeviktor** [reply]
This stubs file is highly explosive 💣 can cause big damage 💥

```php
class Request
{
    /**
     * @return ($arg1 is 'company' ? Company : (next condition here))
     */
    public function route($arg1, ....
}
```
https://phpstan.org/blog/phpstan-1-6-0-with-conditional-return-types
